cmd/generator/main.go

```diff
@@ -213,39 +235,73 @@ func getReleasePlugin(ctx context.Context, client *github.Client, repositoryName
 			return nil, errors.Wrapf(err, "failed to read gzipped plugin bundle for release %s", releaseName)
 		}
 
-		bundleReader := tar.NewReader(gzBundleReader)
-		for {
-			hdr, err := bundleReader.Next()
-			if err == io.EOF {
-				break
-			}
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed to read plugin bundle for release %s", releaseName)
-			}
+		bundleData, err := ioutil.ReadAll(gzBundleReader)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read plugin bundle for release %s", releaseName)
+		}
+
+		manifestData, err := getTarFile(tar.NewReader(bytes.NewReader(bundleData)), "plugin.json")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read manifest from plugin bundle for release %s", releaseName)
+		}
+		plugin.Manifest = mattermostModel.ManifestFromJson(bytes.NewReader(manifestData))
 
-			if path.Base(hdr.Name) != "plugin.json" {
-				continue
+		if plugin.Manifest.IconPath != "" {
+			iconData, err := getTarFile(tar.NewReader(bytes.NewReader(bundleData)), plugin.Manifest.IconPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to read icon data from plugin bundle for release %s", releaseName)
 			}
-			manifest := mattermostModel.ManifestFromJson(bundleReader)
 
-			plugin.Manifest = manifest
-			break
+			logger.Debugf("using icon specified in manifest as %s", plugin.Manifest.IconPath)
+			plugin.IconData = fmt.Sprintf("data:image/svg+xml;base64,%s", base64.StdEncoding.EncodeToString(iconData))
 		}
 	} else {
 		logger.Debugf("skipping download since found existing plugin")
 	}
 
+	if plugin.Manifest == nil {
+		return nil, fmt.Errorf("failed to find plugin manifest for release %s", releaseName)
+	}
+
 	// Reset fields, even if we found the existing plugin above.
-	plugin.HomepageURL = repository.GetHTMLURL()
-	plugin.IconData = ""
+	if plugin.Manifest.HomepageURL != "" {
+		plugin.HomepageURL = plugin.Manifest.HomepageURL
+	} else {
+		plugin.HomepageURL = repository.GetHTMLURL()
+	}
```

Comment:
You forgot `SupportURL`. We can use https://godoc.org/github.com/google/go-github/github#Repository.GetIssuesURL for this

Reply:
Actually, I omitted this on purpose, since it will be captured in the plugin's manifest (when all the plugins are updated). I only "promoted" `HomepageURL` since it's already a field v5.16/v5.17 expects to have at the top-level, outside the manifest.

Perhaps in a `v2` API endpoint, we can strip even this duplication and effectively just deliver manifests.

Reply:
Got it.  :+1: I didn't know `HomepageURL` was a thing before 5.18.